quickwit-metastore/src/metastore/single_file_metastore/store_operations.rs

```diff
@@ -0,0 +1,151 @@
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::path::{Path, PathBuf};
+
+use quickwit_storage::{Storage, StorageError, StorageErrorKind};
+
+use crate::metastore::single_file_metastore::metadata_set::MetadataSet;
+use crate::{MetastoreError, MetastoreResult};
+
+/// Metadata file managed by [`SingleFileMetastore`].
+const META_FILENAME: &str = "quickwit.json";
+
+/// Path to the metadata file from the given index ID.
+pub(crate) fn meta_path(index_id: &str) -> PathBuf {
+    Path::new(index_id).join(Path::new(META_FILENAME))
+}
+
+fn convert_error(index_id: &str, storage_err: StorageError) -> MetastoreError {
+    match storage_err.kind() {
+        StorageErrorKind::DoesNotExist => MetastoreError::IndexDoesNotExist {
+            index_id: index_id.to_string(),
+        },
+        StorageErrorKind::Unauthorized => MetastoreError::Forbidden {
+            message: "The request credentials do not allow for this operation.".to_string(),
+        },
+        _ => MetastoreError::InternalError {
+            message: "Failed to get index files.".to_string(),
+            cause: anyhow::anyhow!(storage_err),
+        },
+    }
+}
+
+pub(crate) async fn fetch_metadata_set(
+    storage: &dyn Storage,
+    index_id: &str,
+) -> MetastoreResult<MetadataSet> {
+    let metadata_path = meta_path(index_id);
+    let content = storage
+        .get_all(&metadata_path)
+        .await
+        .map_err(|storage_err| convert_error(index_id, storage_err))?;
+
+    let metadata_set: MetadataSet = serde_json::from_slice(&content[..])
+        .map_err(|serde_err| MetastoreError::InvalidManifest { cause: serde_err })?;
+
+    if metadata_set.index_id() != index_id {
+        return Err(MetastoreError::InternalError {
+            message: "Inconsistent manifest: index_id mismatch.".to_string(),
+            cause: anyhow::anyhow!(
+                "Expected index_id `{}`, but found `{}`",
+                index_id,
+                metadata_set.index_id()
+            ),
+        });
+    }
+    Ok(metadata_set)
+}
+
+pub(crate) async fn index_exists(storage: &dyn Storage, index_id: &str) -> MetastoreResult<bool> {
+    let metadata_path = meta_path(index_id);
+    let exists = storage
+        .exists(&metadata_path)
+        .await
+        .map_err(|storage_err| convert_error(index_id, storage_err))?;
+    Ok(exists)
+}
+
+/// Serializes the metadata set and stores the data on the storage.
+pub(crate) async fn put_metadata_set_to_path(
```

Comment:
Nitpick: since we pass index_id & make the path inside
```suggestion
pub(crate) async fn put_metadata_set(
```